networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -372,7 +371,290 @@ def asadpour_tsp(G, weight="weight"):
+
+        Returns
+        -------
+        Set
+            The set of minimum 1-Arborescences
+        """
+        # Create a copy of G without vertex 1.
+        G_1 = G.copy()
+        minimum_1_arborescences = set()
+        minimum_1_arborescence_weight = math.inf
+
+        # node is node '1' in the Held and Karp paper
+        n = next(G.__iter__())
+        G_1.remove_node(n)
+
+        # Iterate over the spanning arborescences of the graph until we know
+        # that we have found the minimum 1-arborescences. My proposed strategy
+        # is to find the most extensive root to connect to from 'node 1' and
+        # the least expensive one. We then iterate over arborescences until
+        # the cost of the basic arborescence is the cost of the minimum one
+        # plus the difference between the most and least expensive roots,
+        # that way the cost of connecting 'node 1' will by definition not by
+        # minimum
+        min_root = {"node": None, "weight": math.inf}
+        max_root = {"node": None, "weight": -math.inf}
+        for u, v, d in G.edges(n, data=True):
+            if d[weight] < min_root["weight"]:
+                min_root = {"node": v, "weight": d[weight]}
+            if d[weight] > max_root["weight"]:
+                max_root = {"node": v, "weight": d[weight]}
+
+        min_in_edge = min(G.in_edges(n, data=True), key=lambda x: x[2][weight])
+        min_root["weight"] = min_root["weight"] + min_in_edge[2][weight]
+        max_root["weight"] = max_root["weight"] + min_in_edge[2][weight]
+
+        min_arb_weight = math.inf
+        for arb in nx.ArborescenceIterator(G_1):
+            # Only iterate over the minimum weight arborescences
+            arb_weight = arb.size(weight)
+            if min_arb_weight == math.inf:
+                min_arb_weight = arb_weight
+            elif arb_weight > min_arb_weight + max_root["weight"] - min_root["weight"]:
+                break
+            # We have to pick the root node of the arborescence for the out
+            # edge of the first vertex as that is the only node without an
+            # edge directed into it.
+            for N in arb:
+                if arb.in_degree(N) == 0:
+                    # root found
+                    arb.add_edge(n, N, weight=G[n][N][weight])
+                    arb_weight += G[n][N][weight]
+                    break
+            # We can pick the minimum weight in-edge for the vertex with a
+            # cycle
+            min_in_edge_weight = math.inf
+            min_in_edge = None
+            for u, v, d in G.in_edges(n, data=True):
+                edge_weight = d[weight]
+                if u == N:
+                    continue
+                if edge_weight < min_in_edge_weight:
+                    min_in_edge_weight = edge_weight
+                    min_in_edge = (u, v)
+            arb.add_edge(min_in_edge[0], min_in_edge[1], weight=min_in_edge_weight)
```

Comment:
This call to `add_edge` might be bugged if `weight != "weight"`.  Be sure to test a graph whose weights are stored in a property with a different name.

Reply:
Great catch, I didn't even notice that I used that assumption about the weight of the edges. I've replaced those with an unpacked dict but will definitely still test with a different value for `weight`.